### PR TITLE
Allow optional API methods with no authentication

### LINKED ISSUE #2474: Enable unauthenticated access to specific API methods
### Description
Currently, enabling authentication forces all API Methods to be authenticated. It would be nice that enabling authentication for specific API methods, like ADMIN, was a feature. For example, calls to the more harmless methods like NET or ETH should not need authentication, however, calls to ADMIN methods should be authenticated.

### PR DESCRIPTION
**Pull request by frankisawesome**
## PR description
Adds no auth api methods start up option for some api methods to require no authentication when RPC is configured with authentication

Main changes:
- Add CLI options --rpc-http-api-methods-no-auth and --rpc-ws-api-methods-no-auth and relevant regression tests
- Rename HttpServiceLoginAcceptanceTest to JsonRpcHttpAuthenticationAcceptanceTest as the acceptance test file for all RPC authentication tests
- When no auth api methods are specified no token check is performed for incoming requests on the method, invalid and empty JWT tokens will be allowed

## Fixed Issue(s)

fixes #2474

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcConfiguration.java [hunk 1]
```diff
@@ -41,6 +41,7 @@
   private String host;
   private List<String> corsAllowedDomains = Collections.emptyList();
   private List<String> rpcApis;
+  private List<String> noAtuhRpcApis = Collections.emptyList();
```
Comment (jflo) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcConfiguration.java:44]: typo in member

**Review by jflo — COMMENTED** [review]
I'm curious as to why we add an additional peer to wait for in the tests.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcConfiguration.java (hunk 1 above)
Comment (frankisawesome) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcConfiguration.java]: [db39e60](https://github.com/hyperledger/besu/pull/3382/commits/db39e60b1fc916904e18b67979b55bb7f9977942)

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 2]
```diff
@@ -1922,6 +1963,8 @@ private JsonRpcConfiguration jsonRpcConfiguration() {
     jsonRpcConfiguration.setMaxActiveConnections(rpcHttpMaxConnections);
     jsonRpcConfiguration.setCorsAllowedDomains(rpcHttpCorsAllowedOrigins);
     jsonRpcConfiguration.setRpcApis(rpcHttpApis.stream().distinct().collect(Collectors.toList()));
+    jsonRpcConfiguration.setNoAtuhRpcApis(
```
[the remark this answers is not in the document]
Comment (frankisawesome) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: [db39e60](https://github.com/hyperledger/besu/pull/3382/commits/db39e60b1fc916904e18b67979b55bb7f9977942)

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/jsonrpc/JsonRpcHttpAuthenticationAcceptanceTest.java [hunk 3]
```diff
@@ -113,7 +166,7 @@ public void externalRsaJwtPublicKeyUsedOnJsonRpcMethodShouldFailOnNonPermittedMe
   public void externalEcdsaJwtPublicKeyUsedOnJsonRpcMethodShouldSucceed() {
     nodeUsingEcdsaJwtPublicKey.useAuthenticationTokenInHeaderForJsonRpc(
         ECDSA_TOKEN_ALLOWING_NET_PEER_COUNT);
-    nodeUsingEcdsaJwtPublicKey.verify(net.awaitPeerCount(2));
+    nodeUsingEcdsaJwtPublicKey.verify(net.awaitPeerCount(3));
```
[the remark this answers is not in the document]
Comment (frankisawesome) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/acceptance/jsonrpc/JsonRpcHttpAuthenticationAcceptanceTest.java:168]: I've added a new node that's now running in the cluster so each node should have 3 peers

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 4]
```diff
@@ -569,6 +570,15 @@ void setBannedNodeIds(final List<String> values) {
           "Comma separated list of APIs to enable on JSON-RPC HTTP service (default: ${DEFAULT-VALUE})")
   private final List<String> rpcHttpApis = DEFAULT_RPC_APIS;
 
+  @Option(
+      names = {"--rpc-http-api-method-no-auth", "--rpc-http-api-methods-no-auth"},
```
Comment (pinges) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:574]: I think just having "--rpc-http-api-methods-no-auth" is enough. I find it confusing to have both here ...

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 5]
```diff
@@ -702,6 +712,15 @@ void setBannedNodeIds(final List<String> values) {
           "Comma separated list of APIs to enable on JSON-RPC WebSocket service (default: ${DEFAULT-VALUE})")
   private final List<String> rpcWsApis = DEFAULT_RPC_APIS;
 
+  @Option(
+      names = {"--rpc-ws-api-methods-no-auth", "--rpc-ws-api-method-no-auth"},
```
Comment (pinges) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:716]: s.a.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java [hunk 6]
```diff
@@ -34,6 +36,15 @@ public static boolean isPermitted(
       final Optional<AuthenticationService> authenticationService,
       final Optional<User> optionalUser,
       final JsonRpcMethod jsonRpcMethod) {
+    return isPermitted(authenticationService, optionalUser, jsonRpcMethod, null);
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java:39]: should this method still exist? I thought that only the method with the no auth method should exist ....

**Review by pinges — COMMENTED** [review]
Please have a look at my comments

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 7]
```diff
@@ -1647,6 +1666,28 @@ public void validateRpcOptionsParams() {
     if (!rpcWsApis.stream().allMatch(configuredApis)) {
       throw new ParameterException(this.commandLine, "Invalid value for option '--rpc-ws-apis'");
     }
+
+    final boolean validHttpApiMethods =
+        rpcHttpApiMethodsNoAuth.stream()
+            .map(RpcMethod::rpcMethodExists)
+            .reduce(true, (a, c) -> a && c);
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1673]:
```suggestion
    final boolean validHttpApiMethods =
        rpcHttpApiMethodsNoAuth.stream()
            .allMatch(RpcMethod::rpcMethodExists);
```

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 8]
```diff
@@ -1647,6 +1666,28 @@ public void validateRpcOptionsParams() {
...
+    if (!validHttpApiMethods) {
+      throw new ParameterException(
+          this.commandLine,
+          "Invalid value for option '--rpc-http-api-methods-no-auth', options must be valid RPC methods");
+    }
+
+    final boolean validWsApiMethods =
+        rpcWsApiMethodsNoAuth.stream()
+            .map(RpcMethod::rpcMethodExists)
+            .reduce(true, (a, c) -> a && c);
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1684]:
```suggestion
    final boolean validWsApiMethods =
        rpcWsApiMethodsNoAuth.stream()
            .allMatch(RpcMethod::rpcMethodExists);
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java [hunk 9]
```diff
@@ -42,6 +53,12 @@ public static boolean isPermitted(
       return true;
     }
 
+    // if the method is configured as a no auth method we skip permission check
+    if (Optional.ofNullable(noAuthMethods).orElse(Collections.emptyList()).stream()
```
Comment (siladu) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java:57]: Can noAuthMethods just be an emptyList in all cases?

Comment (frankisawesome) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java:57]: Makes a lot more sense than what I'm doing here!

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java (hunk 6 above)
Comment (frankisawesome) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java:39]: Good point, you made me realise that I didn't implement the auth skipping for web sockets!

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java (hunk 9 above)
Comment (frankisawesome) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java]: 7d80dd7

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 8 above)
Comment (frankisawesome) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: 7d80dd7

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 7 above)
Comment (frankisawesome) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: 7d80dd7

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 4 above)
Comment (frankisawesome) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:574]: ah this is just following the way we did '--rpc-http-api-method' and '--rpc-http-api-methods', keeping it consistent

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/configuration/BesuNodeConfigurationBuilder.java [hunk 10]
```diff
@@ -169,6 +169,18 @@ public BesuNodeConfigurationBuilder jsonRpcAuthenticationConfiguration(final Str
     return this;
   }
 
+  public BesuNodeConfigurationBuilder jsonRpcAuthenticationConfiguration(
+      final String authFile, final List<String> apiMethods) throws URISyntaxException {
```
[the remark this answers is not in the document]
Comment (frankisawesome) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/node/configuration/BesuNodeConfigurationBuilder.java]: 7d80dd7

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketServiceLoginTest.java [hunk 11]
```diff
@@ -144,67 +237,67 @@ public void loginWithBadCredentials(final TestContext context) {
   @Test
   public void loginWithGoodCredentials(final TestContext context) {
     final Async async = context.async();
+    Handler<AsyncResult<HttpClientResponse>> responseHandler =
```
Comment (frankisawesome) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketServiceLoginTest.java:240]: This is a style refactor as the previous layout had too many tabs in

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java (hunk 6 above)
Comment (frankisawesome) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/AuthenticationUtils.java]: [d4e4e92](https://github.com/hyperledger/besu/pull/3382/commits/d4e4e9283a0728fdf5e50439d1132bca4dea2aac)

**Review by pinges — CHANGES_REQUESTED** [review]

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 12]
```diff
@@ -2110,6 +2153,8 @@ private WebSocketConfiguration webSocketConfiguration() {
     webSocketConfiguration.setMaxFrameSize(rpcWsMaxFrameSize);
     webSocketConfiguration.setMaxActiveConnections(rpcWsMaxConnections);
     webSocketConfiguration.setRpcApis(rpcWsApis);
+    webSocketConfiguration.setRpcApisNoAuth(
+        rpcHttpApiMethodsNoAuth.stream().distinct().collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (frankisawesome) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2157]: good catch!

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java [hunk 13]
```diff
@@ -1941,6 +1942,48 @@ public void rpcApisPropertyIgnoresDuplicatesAndMustBeUsed() {
...
+  @Test
+  public void rpcHttpNoAuthApiMethodsCannotBeInvalid() {
+    parseCommand("--rpc-http-enabled", "--rpc-http-api-method-no-auth", "invalid");
+
+    Mockito.verifyNoInteractions(mockRunnerBuilder);
+
+    assertThat(commandOutput.toString(UTF_8)).isEmpty();
+    assertThat(commandErrorOutput.toString(UTF_8))
+        .contains(
+            "Invalid value for option '--rpc-http-api-methods-no-auth', options must be valid RPC methods");
```
[the remark this answers is not in the document]
Comment (macfarla) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1972]: there is a separate issue open for this - regular rpc-http-api methods have the same issue

[the remark this answers is not in the document]
Comment (macfarla) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1972]: https://github.com/hyperledger/besu/issues/3418

[the remark this answers is not in the document]
Comment (macfarla) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1972]: meaning - I would leave that as a separate change PR

[the remark this answers is not in the document]
Comment (frankisawesome) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1972]: I'll do a seperate PR for both

**Review by pinges — APPROVED** [review]

**Pull request merged**